x/upgrade/keeper/msg_server.go

```diff
@@ -0,0 +1,62 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+	gov "github.com/cosmos/cosmos-sdk/x/gov/types"
+	"github.com/cosmos/cosmos-sdk/x/upgrade/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the upgrade MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(k Keeper) types.MsgServer {
+	return &msgServer{
+		Keeper: k,
+	}
+}
+
+var _ types.MsgServer = msgServer{}
+
+// SoftwareUpgrade implements the Msg/SoftwareUpgrade Msg service.
+func (k msgServer) SoftwareUpgrade(goCtx context.Context, req *types.MsgSoftwareUpgrade) (*types.MsgSoftwareUpgradeResponse, error) {
+	govAcct := k.authKeeper.GetModuleAddress(gov.ModuleName).String()
+	if govAcct != req.Authority {
```

Comment:
```suggestion
	if k.authority != req.Authority {
```

same below in Cancel

Reply:
Note: with the above suggested change, as long as we keep gov@v1beta1, there's always a way to bypass the `k.authority` check, simply by sending a legacy upgrade proposal. The legacy one will be performed by gov. This is maybe fine though.

Reply:
Is it possible that an application disable that behaviour whereby a legacy upgrade proposal is always executed by the gov module. I guess they could just remove the handler right?

Reply:
Yes, true they can.

Another solution we can do is: add the same `authority` field in the gov keeper. Then it's the app role to make sure that both keepers' authorities are the same.